presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ProjectOffPushDownRule.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Pattern;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.iterative.rule.Util.pruneInputs;
+
+/**
+ * @param <N> The node type to look for under the ProjectNode
+ * Looks for a Project parent over a N child, such that the parent doesn't use all the output columns of the child.
+ * Given that situation, invokes the pushDownProjectOff helper to possibly rewrite the child to produce fewer outputs.
+ */
+public abstract class ProjectOffPushDownRule<N extends PlanNode>
```

Comment:
My understanding of things:
* "prune columns" is the overall goal of all of the PruneFooColumns rules, working together
* "project off pushdown" is the algorithm the rules are using to collectively achieve that goal
* "restrict outputs" is a helper operation that is useful in other situations as well.  Changing its name to be related to "project off" would be nice, though the most convenient form for it takes the preserved columns as the argument, leading to an awkward name like "projectOffAnythingOtherThan".

Renaming the rules to all be something about pushing down project-off would make me happy, since that's what they're doing.  I was just going with Martin's existing naming.